File: benchmarks/problem12/go/main.go
package main

import (
	"fmt"
	"strconv"
)

var _ = fmt.Printf
var _ = strconv.FormatInt

type PrimeGenerator struct {
	n, last int64
	sieve   map[int64]int64
}

func MakePrimeGenerator() *PrimeGenerator {
	return &PrimeGenerator{3, 2, make(map[int64]int64)}
}

func (self *PrimeGenerator) Next() int64 {
	prime, ok := self.sieve[self.n]
	for ok {
		delete(self.sieve, self.n)

		var composite int64 = self.n + prime + prime
		_, ok_ := self.sieve[composite]
		for ok_ {
			composite += prime + prime
			_, ok_ = self.sieve[composite]
		}
		self.sieve[composite] = prime

		self.n += 2
		prime, ok = self.sieve[self.n]
	}

	self.sieve[self.n*self.n] = self.n
	r := self.last
	self.last = self.n
	self.n += 2
	return r
}

type FactorFinder struct {
	known       map[int64][]int64
	nextPrimes  *PrimeGenerator
	knownPrimes []int64
}

func MakeFactorFinder() *FactorFinder {
	known := make(map[int64][]int64)
	known[1] = []int64{1}
	return &FactorFinder{known, MakePrimeGenerator(), make([]int64, 0)}
}

func (self *FactorFinder) GetPrimeFactors(of int64) []int64 {
	factors := make([]int64, 0)
	quotient := of

	for _, prime := range self.knownPrimes {
		if prime > quotient {
			return factors
		}

		remainder := quotient % prime
		for remainder == 0 {
			quotient = quotient / prime
			remainder = quotient % prime
			factors = append(factors, prime)
		}
	}

	for {
		prime := self.nextPrimes.Next()
		self.knownPrimes = append(self.knownPrimes, prime)
		if prime > quotient {
			return factors
		}

		remainder := quotient % prime
		for remainder == 0 {
			quotient = quotient / prime
			remainder = quotient % prime
			factors = append(factors, prime)
		}
	}

	return factors
}

func (self *FactorFinder) GetFactors(of int64) []int64 {

	if facs, ok := self.known[of]; ok {
		return facs
	}

	primeFactors := self.GetPrimeFactors(of)
	factorSet := make(map[int64]bool)
	factorSet[1] = true
	factorSet[of] = true

	for _, prime := range primeFactors {

		factor := of / prime
		for _, subfactor := range self.GetFactors(factor) {
			factorSet[subfactor] = true
		}
	}

	keys := make([]int64, len(factorSet))
	i := 0
	for k := range factorSet {
		keys[i] = k
		i++
	}

	self.known[of] = keys
	return keys
}

func Solve() int64 {

	finder := MakeFactorFinder()
	var adder int64 = 0
	var tn int64 = 0

	for {
		adder += 1
		tn += adder
		tnFactors := finder.GetFactors(tn)

		if len(tnFactors) > 1000 {
			return tn
		}
	}
}

func main() {
	answer := Solve()
	println("Answer: ", answer)
}

File: benchmarks/problem12/fsharp/main.fs
open System.Collections.Generic

type PrimeGenerator =
    val mutable n: int64
    val mutable last: int64
    val mutable sieve: Dictionary<int64, int64>

    new() = { n = 3L; last = 2L; sieve = Dictionary<int64, int64>() }

    member this.Next() =
        let mutable prime: int64 = 0L
        while this.sieve.TryGetValue(this.n, &prime) do
            this.sieve.Remove(this.n) |> ignore;

            let mutable composite = this.n + prime + prime
            while this.sieve.ContainsKey(composite) do
                composite <- composite + prime + prime
            this.sieve.Add(composite, prime)

            this.n <- this.n + 2L

        this.sieve.Add(this.n * this.n, this.n)
        let r = this.last
        this.last <- this.n
        this.n <- this.n + 2L
        r

type FactorFinder =
    val known: Dictionary<int64, HashSet<int64>>
    val nextPrimes: PrimeGenerator
    val knownPrimes: List<int64>

    new() = {
        known = (let k = Dictionary<int64, HashSet<int64>>()
                 let initKnown = HashSet<int64>()
                 initKnown.Add(1L) |> ignore
                 k.[1L] <- initKnown
                 k);
        nextPrimes = PrimeGenerator();
        knownPrimes = List<int64>();

    }

    member this.GetPrimeFactors(ofn: int64): List<int64> =
        let factors = List<int64>()
        let mutable quotient = ofn
        let mutable kprimes = this.knownPrimes.GetEnumerator()

        while kprimes.MoveNext() && kprimes.Current <= quotient do
            let mutable remainder = quotient % kprimes.Current
            while remainder = 0L do
                quotient <- quotient / kprimes.Current
                remainder <- quotient % kprimes.Current
                factors.Add(kprimes.Current)

        let mutable prime: int64 = this.nextPrimes.Next()
        this.knownPrimes.Add(prime)
        while prime <= quotient do

            let mutable remainder = quotient % prime
            while remainder = 0L do
                quotient <- quotient / prime
                remainder <- quotient % prime
                factors.Add(prime)

            prime <- this.nextPrimes.Next()
            this.knownPrimes.Add(prime)

        factors

    member this.GetFactors(ofn: int64): HashSet<int64> =
        match this.known.TryGetValue(ofn) with
        | (true, facs) -> facs
        | (false, _) ->
            let primeFactors = this.GetPrimeFactors(ofn)
            let factors = HashSet<int64>()
            factors.Add(1L) |> ignore;
            factors.Add(ofn) |> ignore;
            for prime in primeFactors do
                let factor = ofn / prime
                for subfactor in this.GetFactors(factor) do
                    factors.Add(subfactor) |> ignore;

            this.known.Add(ofn, factors)
            factors

let solve(): int64 =
    let finder = FactorFinder()
    let mutable adder = 0L
    let mutable tn = 0L
    let mutable factorCount = 0

    while factorCount < 1000 do
        adder <- adder + 1L
        tn <- tn + adder
        let tnFactors = finder.GetFactors tn
        let count = tnFactors.Count
        factorCount <- count

    tn

[<EntryPoint>]
let main args =
    let answer: int64 = solve()
    printfn "Answer: %i" answer
    0